toolkit/tools/pkg/imagecustomizerlib/rpmsourcesmounts.go

```diff
@@ -0,0 +1,338 @@
+				}
+
+				// Change the baseurl to point to the bind mount directory.
+				newBaseurl := fmt.Sprintf("file://%s", mountTargetDirectoryInChroot)
+				repoConfig.Key("baseurl").SetValue(newBaseurl)
+			}
+		}
+
+		// Copy over the repo details to the all-repos config.
+		err := appendIniSection(allReposConfig, repoConfig)
+		if err != nil {
+			return fmt.Errorf("failed to append repo config:\n%w", err)
+		}
+	}
+
+	return nil
+}
+
+func (m *rpmSourcesMounts) mountRpmsDirectory(rpmSourceName string, rpmsDirectory string,
+	imageChroot *safechroot.Chroot,
+) (string, error) {
+	i := len(m.mounts)
+	targetName := fmt.Sprintf("%02d%s", i, rpmSourceName)
+	mountTargetDirectoryInChroot := path.Join(rpmsMountParentDirInChroot, targetName)
+	mountTargetDirectory := path.Join(imageChroot.RootDir(), mountTargetDirectoryInChroot)
+
+	// Create a read-only bind mount.
+	mount, err := safemount.NewMount(rpmsDirectory, mountTargetDirectory, "", unix.MS_BIND|unix.MS_RDONLY, "", true)
+	if err != nil {
+		return "", fmt.Errorf("failed to mount RPM source directory from (%s) to (%s):\n%w", rpmsDirectory, mountTargetDirectory, err)
+	}
+
+	m.mounts = append(m.mounts, mount)
+	return mountTargetDirectoryInChroot, nil
+}
+
+func (m *rpmSourcesMounts) close() error {
+	var err error
+	var errs []error
+
+	// Delete allrepos.repo file (if it exists).
+	err = os.RemoveAll(m.allReposConfigFilePath)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	// Unmount rpm source directories.
+	for _, mount := range m.mounts {
+		err = mount.Close()
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+	}
+
+	// Join all the errors together.
+	if len(errs) > 0 {
+		err = errors.Join(errs...)
+		err = fmt.Errorf("failed to cleanup RPM sources mounts:\n%w", err)
+		return err
+	}
+
+	// Delete the temporary directory.
+	if m.rpmsMountParentDirCreated {
+		// Note: Do not use `RemoveAll` here in case there are any leftover mounts that failed to unmount.
+		err = os.Remove(m.rpmsMountParentDir)
+		if err != nil {
+			return fmt.Errorf("failed to delete source rpms directory (%s):\n%w", m.rpmsMountParentDir, err)
+		}
+
+		m.rpmsMountParentDirCreated = false
+	}
+
+	return nil
+}
+
+func getRpmSourceFileType(rpmSourcePath string) (string, error) {
+	// First, check if path points to a directory.
+	isDir, err := file.IsDir(rpmSourcePath)
+	if err != nil {
+		return "", fmt.Errorf("failed to get type of RPM source (%s):\n%w", rpmSourcePath, err)
+	}
+
+	if isDir {
+		return "dir", nil
+	}
+
+	filename := filepath.Base(rpmSourcePath)
+	dotIndex := strings.Index(filename, ".")
+	fileExt := ""
+	if dotIndex >= 0 {
+		fileExt = filename[dotIndex:]
+	}
+
+	switch fileExt {
+	case ".conf":
+		return "conf", nil
+
+	default:
+		return "", nil
+	}
+}
+
+// Add a local directory containing RPMs to the allrepos.repo file.
+func appendLocalRepo(iniFile *ini.File, mountTargetDirectoryInChroot string) error {
```

Comment:
side question - are we planning on supporting external repos through an API?

Reply:
If you pass in an RPM repo config file as an RPM source, you can point that to an external repo.